### PR TITLE
Fixed crash on process exit on Windows

### PR DESCRIPTION
**Pull request by alespergl**
This is a fix for #10188 which was introduced by switching to dynamically linked CRT on Windows. The problem went unnoticed because it is mostly harmless other than recording the app crash in the Windows Event Log. It is desirable to merge this to master and to the 1.7 stream as well.

### Description:

Chromium has its own TLS subsystem which supports automatic destruction of thread-local data, and also depends on memory allocation routines provided by the CRT. The problem is that the auto-destruction mechanism uses a hidden feature of the OS loader which calls a callback on thread exit, but only after all loaded DLLs have been detached. Since the CRT is also a DLL, it happens that by the time Chromium's `OnThreadExit` function is called, the heap functions, though still in memory, no longer perform their duties, and when Chromium calls `free` on its buffer, it triggers an access violation error. 

The workaround is to invoke Chromium's `OnThreadExit` in time from within the CRT's atexit facility, ensuring the heap functions are still active. The second invocation from the OS loader code will be a no-op.

### CODE REVIEW

**Comment by MarshallOfSound** [comment]
Awesome, let's get this in asap :star:

File: atom/app/atom_main.cc [hunk 1]
```diff
@@ -94,6 +94,23 @@ int APIENTRY wWinMain(HINSTANCE instance, HINSTANCE, wchar_t* cmd, int) {
...
+  // of thread-local data, and also depends on memory allocation routines
+  // provided by the CRT. The problem is that the auto-destruction mechanism
+  // uses a hidden feature of the OS loader which calls a callback on thread
+  // exit, but only after all loaded DLLs have been detached. Since the CRT is
+  // also a DLL, it happens that by the time Chromium's `OnThreadExit` function
+  // is called, the heap functions, though still in memory, no longer perform
+  // their duties, and when Chromium calls `free` on its buffer, it triggers
+  // an access violation error.
+  // We work around this problem by invoking Chromium's `OnThreadExit` in time
+  // from within the CRT's atexit facility, ensuring the heap functions are still
```
Comment (zcbenz) [inline atom/app/atom_main.cc]: It is triggering linter error:

    atom/app/atom_main.cc:107:  Lines should be <= 80 characters long  [whitespace/line_length] [2]

**Comment by alespergl** [comment]
Hm, seems like something changed in Chromium since 58 and the OnThreadExit function is not there anymore.

Can we get this into 1.7 as is (there's no branch for that as far as I can tell) while I figure out how to port it to 1.8?

**Comment by jkleinsc** [comment]
@alespergl There is a 1.7.x branch here: https://github.com/electron/electron/tree/1-7-x

**Comment by jkleinsc** [comment]
@alespergl thanks for updating the PR.  It looks like the Windows 32 bit build is failing:
```
App threw an error during load
Error: Cannot find module 'electabul'
    at Module._resolveFilename (module.js:470:15)
    at Function.Module._resolveFilename (C:\e\workspace\electron-win-ia32\out\D\resources\electron.asar\common\reset-search-paths.js:35:12)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (C:\e\workspace\electron-win-ia32\spec\static\main.js:7:20)
    at Object.<anonymous> (C:\e\workspace\electron-win-ia32\spec\static\main.js:377:3)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
```

Linux x64 build fails with the following:
```
not ok 111 BrowserWindow module "webPreferences" option nativeWindowOpen option loads native addons correctly after reload
  AssertionError: 'Require runas failed: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.21\' not found (required by /var/lib/jenki == 'function'
      at ipcMain.once (/var/lib/jenkins/workspace/electron-linux-x64/spec/api-browser-window-spec.js:1359:18)
      at CallbacksRegistry.apply (/var/lib/jenkins/workspace/electron-linux-x64/out/D/resources/electron.asar/common/api/callbacks-registry.js:48:25)
      at EventEmitter.<anonymous> (/var/lib/jenkins/workspace/electron-linux-x64/out/D/resources/electron.asar/renderer/api/remote.js:299:21)
      at emitThree (events.js:116:13)
      at EventEmitter.emit (events.js:197:7)
```
I am going to enable CircleCI on this branch to see if running the Linux x64 build will work there.

**Comment by jkleinsc** [comment]
@alespergl All builds are now passing, so I think this is good to go.

**Pull request merged**